// @lay2/pw-core/build/main/providers/raw-provider.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RawProvider = void 0;
const provider_1 = require("./provider");
const models_1 = require("../models");
const ckb_sdk_utils_1 = require("@nervosnetwork/ckb-sdk-utils");
const hashers_1 = require("../hashers");
const ecpair_1 = __importDefault(require("@nervosnetwork/ckb-sdk-utils/lib/ecpair"));
class RawProvider extends provider_1.Provider {
    constructor(privateKey) {
        super(provider_1.Platform.ckb);
        this.privateKey = privateKey;
        this.keyPair = new ecpair_1.default(privateKey);
    }
    async init() {
        const pwPrefix = models_1.getDefaultPrefix();
        const prefix = pwPrefix === models_1.AddressPrefix.ckb
            ? ckb_sdk_utils_1.AddressPrefix.Mainnet
            : ckb_sdk_utils_1.AddressPrefix.Testnet;
        const address = ckb_sdk_utils_1.privateKeyToAddress(this.privateKey, { prefix });
        this.address = new models_1.Address(address, models_1.AddressType.ckb);
        return this;
    }
    hasher() {
        return new hashers_1.Blake2bHasher();
    }
    async sign(message) {
        const sig = this.keyPair.signRecoverable(message);
        return sig;
    }
    async close() {
        return true;
    }
}
exports.RawProvider = RawProvider;
