// @lay2/pw-core/build/module/providers/raw-provider.js
import { Platform, Provider } from './provider';
import { Address, AddressType, getDefaultPrefix, AddressPrefix as PwAddressPrefix, } from '../models';
import { AddressPrefix, privateKeyToAddress, } from '@nervosnetwork/ckb-sdk-utils';
import { Blake2bHasher } from '../hashers';
import ECPair from '@nervosnetwork/ckb-sdk-utils/lib/ecpair';
export class RawProvider extends Provider {
    constructor(privateKey) {
        super(Platform.ckb);
        this.privateKey = privateKey;
        this.keyPair = new ECPair(privateKey);
    }
    async init() {
        const pwPrefix = getDefaultPrefix();
        const prefix = pwPrefix === PwAddressPrefix.ckb
            ? AddressPrefix.Mainnet
            : AddressPrefix.Testnet;
        const address = privateKeyToAddress(this.privateKey, { prefix });
        this.address = new Address(address, AddressType.ckb);
        return this;
    }
    hasher() {
        return new Blake2bHasher();
    }
    async sign(message) {
        const sig = this.keyPair.signRecoverable(message);
        return sig;
    }
    async close() {
        return true;
    }
}
