// @lay2/pw-core/build/main/builders/builder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Builder = void 0;
const models_1 = require("../models");
const __1 = __importDefault(require(".."));
const FEE_BASE = 1000;
class Builder {
    constructor(feeRate = Builder.MIN_FEE_RATE, collector = __1.default.defaultCollector, witnessArgs = Builder.WITNESS_ARGS.Secp256k1) {
        this.feeRate = feeRate;
        this.collector = collector;
        this.witnessArgs = witnessArgs;
    }
    static calcFee(tx, feeRate = Builder.MIN_FEE_RATE) {
        if (feeRate < Builder.MIN_FEE_RATE) {
            feeRate = Builder.MIN_FEE_RATE;
        }
        const txSize = tx.getSize();
        const fee = (feeRate / FEE_BASE) * txSize;
        return new models_1.Amount(fee.toString(), models_1.AmountUnit.shannon);
    }
    getFee() {
        return this.fee;
    }
}
exports.Builder = Builder;
Builder.MIN_FEE_RATE = 1000;
Builder.MIN_CHANGE = new models_1.Amount('61', models_1.AmountUnit.ckb);
Builder.WITNESS_ARGS = {
    Secp256k1: {
        lock: '0x' + '0'.repeat(132),
        input_type: '',
        output_type: '',
    },
    RawSecp256k1: {
        lock: '0x' + '0'.repeat(130),
        input_type: '',
        output_type: '',
    },
    Secp256r1: {
        lock: '0x' + '0'.repeat(600),
        input_type: '',
        output_type: '',
    },
};
