// @nervosnetwork/ckb-sdk-utils/lib/serialization/transaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeTransaction = exports.serializeRawTransaction = exports.serializeWitnesses = exports.serializeWitnessArgs = exports.serializeOutputsData = exports.serializeOutputs = exports.serializeOutput = exports.serializeInputs = exports.serializeInput = exports.serializeHeaderDeps = exports.serializeCellDeps = exports.serializeCellDep = exports.serializeDepType = exports.serializeOutPoint = exports.serializeVersion = void 0;
const script_1 = require("./script");
const convertors_1 = require("../convertors");
const basic_1 = require("./basic");
exports.serializeVersion = (version) => convertors_1.toUint32Le(version);
exports.serializeOutPoint = (outPoint) => {
    if (!outPoint)
        return '';
    const struct = new Map([
        ['txHash', outPoint.txHash],
        ['index', convertors_1.toUint32Le(outPoint.index)],
    ]);
    return basic_1.serializeStruct(struct);
};
exports.serializeDepType = (type) => {
    if (type === 'code')
        return '0x00';
    if (type === 'depGroup')
        return '0x01';
    throw new TypeError("Dep type must be either of 'code' or 'depGroup'");
};
exports.serializeCellDep = (dep) => {
    const serializedOutPoint = exports.serializeOutPoint(dep.outPoint);
    const serializedDepType = exports.serializeDepType(dep.depType);
    const struct = new Map([
        ['outPoint', serializedOutPoint],
        ['depType', serializedDepType],
    ]);
    return basic_1.serializeStruct(struct);
};
exports.serializeCellDeps = (cellDeps) => {
    const serializedCellDepList = cellDeps.map(dep => exports.serializeCellDep(dep));
    return basic_1.serializeFixVec(serializedCellDepList);
};
exports.serializeHeaderDeps = (deps) => {
    const serializedHeaderDepList = deps.map(dep => basic_1.serializeArray(dep));
    return basic_1.serializeFixVec(serializedHeaderDepList);
};
exports.serializeInput = (input) => {
    const serializedOutPoint = exports.serializeOutPoint(input.previousOutput);
    const serializedSince = convertors_1.toUint64Le(input.since);
    const struct = new Map([
        ['since', serializedSince],
        ['previousOutput', serializedOutPoint],
    ]);
    return basic_1.serializeStruct(struct);
};
exports.serializeInputs = (inputs) => {
    const serializedInputList = inputs.map(input => exports.serializeInput(input));
    return basic_1.serializeFixVec(serializedInputList);
};
exports.serializeOutput = (output) => {
    const serializedCapacity = convertors_1.toUint64Le(output.capacity);
    const serializedLockScript = script_1.serializeScript(output.lock);
    const serialiedTypeScript = output.type ? script_1.serializeScript(output.type) : '';
    const table = new Map([
        ['capacity', serializedCapacity],
        ['lock', serializedLockScript],
        ['type', serialiedTypeScript],
    ]);
    return basic_1.serializeTable(table);
};
exports.serializeOutputs = (outputs) => {
    const serializedOutputList = outputs.map(output => exports.serializeOutput(output));
    return basic_1.serializeDynVec(serializedOutputList);
};
exports.serializeOutputsData = (outputsData) => {
    const serializedOutputsDatumList = outputsData.map(datum => basic_1.serializeFixVec(datum));
    return basic_1.serializeDynVec(serializedOutputsDatumList);
};
exports.serializeWitnessArgs = (witnessArgs) => {
    const [serializedLock, serializedInputType, serializedOutputType] = [
        witnessArgs.lock,
        witnessArgs.inputType,
        witnessArgs.outputType,
    ].map(args => {
        if (basic_1.serializeOption(args) === '0x') {
            return '0x';
        }
        return basic_1.serializeFixVec(args);
    });
    const table = new Map([
        ['lock', serializedLock],
        ['inputType', serializedInputType],
        ['outputType', serializedOutputType],
    ]);
    return basic_1.serializeTable(table);
};
exports.serializeWitnesses = (witnesses) => {
    const serializedWitnessList = witnesses.map(witness => basic_1.serializeFixVec(witness));
    return basic_1.serializeDynVec(serializedWitnessList);
};
exports.serializeRawTransaction = (rawTransaction) => {
    const serializedVersion = exports.serializeVersion(rawTransaction.version);
    const serializedCellDeps = exports.serializeCellDeps(rawTransaction.cellDeps);
    const serializedHeaderDeps = exports.serializeHeaderDeps(rawTransaction.headerDeps);
    const serializedInputs = exports.serializeInputs(rawTransaction.inputs);
    const serializedOutputs = exports.serializeOutputs(rawTransaction.outputs);
    const serializedOutputsData = exports.serializeOutputsData(rawTransaction.outputsData);
    const table = new Map([
        ['version', serializedVersion],
        ['cellDeps', serializedCellDeps],
        ['headerDeps', serializedHeaderDeps],
        ['inputs', serializedInputs],
        ['outputs', serializedOutputs],
        ['outputsData', serializedOutputsData],
    ]);
    return basic_1.serializeTable(table);
};
exports.serializeTransaction = (rawTransaction) => {
    const serializedRawTransaction = exports.serializeRawTransaction(rawTransaction);
    const serializedWitnesses = exports.serializeWitnesses(rawTransaction.witnesses || []);
    const table = new Map([
        ['raw', serializedRawTransaction],
        ['witnesses', serializedWitnesses],
    ]);
    return basic_1.serializeTable(table);
};
